### PR TITLE
feat(imagekit-adapter): add CI, tests, error handling

### PR DESCRIPTION
**Pull request by abhinavr4**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,128 +1,156 @@
...
+import got from "got";
+import ImageKit from "imagekit";
+import { join } from "path";
+import serveStatic from "serve-static";
+
+import { readFileAsync, stripLeadingSlash } from "../utils";
+import ImageKitAdapterError from "../utils/errors";
+
+const fileServe = serveStatic(
+  join(process.env.GHOST_CONTENT as string, "images"),
```
Comment (ahnv) [inline src/index.ts:11]: Can we use defaults from [`getContentPath()`](https://github.com/TryGhost/Ghost/blob/7e349d0bf854a1035be1c3c69b69260c42e99e5f/ghost/core/core/shared/config/helpers.js#L69-L95) instead of just relying on GHOST_CONTENT environment variable

File: src/index.ts [hunk 2]
```diff
@@ -1,128 +1,156 @@
...
+  private urlEndpoint: string;
+  private privateKey: string;
+  private publicKey: string;
+  private enableDatedFolders: boolean;
+  private uploadOptions: ImageKitStoreConfig["uploadOptions"];
+  private _imagekit: ImageKit;
+
+  constructor(config: ImageKitStoreConfig) {
+    // @ts-ignore
+    super(config);
```
Comment (ahnv) [inline src/index.ts]: Do we need to send config to `super()` because StorageBase doesn't need config in constructor for [StorageBase](https://github.com/TryGhost/Ghost-Storage-Base/blob/main/BaseStorage.js)

**Review by ahnv — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 2 above)
Comment (abhinavr4) [reply src/index.ts]: done

File: src/index.ts (hunk 1 above)
Comment (abhinavr4) [reply src/index.ts:11]: Added info in docs on setting env variable.

**Review by ahnv — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+- The optional `enableDatedFolders` setting allows uploading images into dated sub-directories (like the default Ghost Local Storage Adapter). It is `true` by default.
+- The `auth` property is used to configure your ImageKit account's credentials and URL endpoint. You can find them in the [dashboard](https://imagekit.io/dashboard/developer/api-keys).
+- The `uploadOptions` property allows you to configure ImageKit's upload options. You can configure `useUniqueFileName`, `folder`, and `tags`.
+
+### Recommended configuration
+
+- `uploadOptions.useUniqueFileName = true` Ghost's local storage adaptor handles duplicate file names automatically. When uploading a file to a location where a file with the same name already exists, ImageKit creates a new file version. In order to avoid this, `useUniqueFileName` can be set to `true`.
+- `uploadOptions.tags = ["travel", "discover"]` if you want to add associations of tags to your uploaded images.
+- `uploadOptions.folder = "/ghost/blog"` allows to upload all your images into a specific directory in your ImageKit media library. By default, files are uploaded to the root of media library.
+
```
Comment (niveditn) [inline README.md:96]: @abhinavr4 Can we add the sample for config file here in the README as well?  It would help the user visualize the config options immediately instead of them having to check the `configuration.json.dist` file first. Simply like this:

```
{
    "storage": {
        "active": "imagekit",
        "imagekit": {
            "enableDatedFolders": true,
            "auth": {
                "privateKey": "your-private-key",
                "publicKey": "your-public-key",
                "urlEndpoint": "https://ik.imagekit.io/your-imagekit-id"
            },
            "uploadOptions": {
                "useUniqueFileName": false,
                "folder": "/sample",
                "tags": ["blog"]
            }
        }
    },
    "imageOptimization": {
        "resize": false
    }
}
```

File: tests/adapter/read.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    } catch (err) {
+      expect(err).to.be.an.instanceOf(ImageKitAdapterError);
+      expect(err.statusCode).to.be.equal(400);
+      expect(err.message).to.be.equal(
+        "Failed to read image from path 'undefined'"
+      );
+    }
+  });
+
+  it("Should return an error for a non-existent file", async function () {
```
Comment (niveditn) [inline tests/adapter/read.ts:39]: Can we add a test for private file?

File: tests/adapter/exists.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+    const exists = await imagekitAdapter.exists(
+      mockImage.name,
+      mockImage.targetDir
+    );
+
+    expect(exists).to.equals(true);
+  });
+
+  it("Should return false when image does not exist", async function () {
```
Comment (niveditn) [inline tests/adapter/exists.ts:23]: Let's add one more for private file.

File: src/index.ts [hunk 6]
```diff
@@ -1,128 +1,159 @@
...
-        return result.toString()
+  private urlEndpoint: string;
+  private privateKey: string;
+  private publicKey: string;
+  private enableDatedFolders: boolean;
+  private uploadOptions: ImageKitStoreConfig["uploadOptions"];
+  private _imagekit: ImageKit;
+
+  constructor(config: ImageKitStoreConfig) {
+    super();
```
Comment (niveditn) [inline src/index.ts:41]: Why was this changed from `super(config)`?

File: src/index.ts [hunk 7]
```diff
@@ -1,128 +1,159 @@
...
-                res.status(404).end();
-            });
-        };
+    return false;
+  }
+
+  async save(image: StorageBase.Image, targetDir?: string): Promise<string> {
+    console.log("save", image.name);
+    try {
+      const folder =
```
Comment (niveditn) [inline src/index.ts:94]: The old logic was
```
const folder = targetDir || this.getTargetDir(this.uploadOptions.folder)
```
In the new one, would `targetDir` not be used directly?

**Review by niveditn — CHANGES_REQUESTED** [review]

File: configuration.json.dist [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    "storage": {
+        "active": "imagekit",
+        "imagekit": {
+            "enableDatedFolders": true,
+            "auth": {
+                "privateKey": "your-private-key",
+                "publicKey": "your-public-key",
+                "urlEndpoint": "https://ik.imagekit.io/your-imagekit-id"
+            },
+            "uploadoptions": {
```
Comment (niveditn) [inline configuration.json.dist]: `uploadOptions`
Please double check all keys in config.

File: src/index.ts (hunk 6 above)
Comment (abhinavr4) [reply src/index.ts:41]: This is because the constructor doesn't have any supported parameters.

File: src/index.ts (hunk 7 above)
Comment (abhinavr4) [reply src/index.ts:94]: In the old logic `this.getTargetDir(this.uploadOptions.folder)` was always called in normal flow. Now we have another flag to support dated folder structure for uploaded images.

If `targetDir` is passed to method then it will be set for folder path.

File: configuration.json.dist (hunk 8 above)
Comment (abhinavr4) [reply configuration.json.dist]: done

File: tests/adapter/exists.ts (hunk 5 above)
Comment (abhinavr4) [reply tests/adapter/exists.ts:23]: Done. Added for signed URL setting also.

File: tests/adapter/read.ts (hunk 4 above)
Comment (abhinavr4) [reply tests/adapter/read.ts:39]: Done. Added for signed URL setting also.

File: README.md (hunk 3 above)
Comment (abhinavr4) [reply README.md:96]: done

**Review by niveditn — APPROVED** [review]
